File: proposed-projects/round-trip-experiment/reverse/20260909T153339Z_9de9fd66/20260910T090338Z_5eab4940/ported_implementation/src/grammar-graph/generic-set.ts
/** A set of elements deduplicated by a derived key, preserving insertion order. */
export class GenericSet<T, K> {
  private keys = new Map<K, T>();
  private set = new Set<T>();

  constructor(private getKey: (el: T) => K) {}

  add(el: T): void {
    const key = this.getKey(el);
    if (!this.keys.has(key)) {
      this.keys.set(key, el);
      this.set.add(el);
    }
  }

  delete(el: T): void {
    const key = this.getKey(el);
    const ref = this.keys.get(key);
    if (ref === undefined) {
      throw new Error('Could not get ref');
    }
    this.keys.delete(key);
    this.set.delete(ref);
  }

  has(el: T): boolean {
    return this.set.has(el);
  }

  get(el: T): T | undefined {
    return this.keys.get(this.getKey(el));
  }

  [Symbol.iterator](): IterableIterator<T> {
    return this.set.values();
  }

  get size(): number {
    return this.set.size;
  }
}

File: packages/gbnf-experiment/data/20260909T153339Z_9de9fd66/ported_implementation/gbnf/grammar_graph/generic_set.py
from typing import Callable, Dict, Generic, Iterator, TypeVar

T = TypeVar("T")
K = TypeVar("K")


class GenericSet(Generic[T, K]):
    """A set of elements deduplicated by a derived key, preserving insertion order."""

    def __init__(self, get_key: Callable[[T], K]):
        self.get_key = get_key
        self._keys: Dict[K, T] = {}
        # keyed by object identity, mirroring a JS Set of objects
        self._set: Dict[int, T] = {}

    def add(self, el: T) -> None:
        key = self.get_key(el)
        if key not in self._keys:
            self._keys[key] = el
            self._set[id(el)] = el

    def delete(self, el: T) -> None:
        key = self.get_key(el)
        ref = self._keys.get(key)
        if ref is None:
            raise Exception("Could not get ref")
        del self._keys[key]
        self._set.pop(id(ref), None)

    def has(self, el: T) -> bool:
        return id(el) in self._set

    def get(self, el: T):
        return self._keys.get(self.get_key(el))

    def __iter__(self) -> Iterator[T]:
        return iter(list(self._set.values()))

    @property
    def size(self) -> int:
        return len(self._set)

    def __len__(self) -> int:
        return len(self._set)
